Source/Windows/WOW64/IntervalList.h

```diff
@@ -0,0 +1,143 @@
+#pragma once
+
+#include <utility>
+#include <algorithm>
+
+#include <FEXCore/fextl/vector.h>
+
+template<typename SizeType>
+class IntervalList {
+public:
+  using DifferenceType = decltype(std::declval<SizeType>() - std::declval<SizeType>());
+
+  struct Interval {
+    SizeType Offset;
+    SizeType End;
+
+    Interval() = default;
+
+    Interval(SizeType Offset, SizeType End) : Offset{Offset}, End{End} {}
+  };
+
+private:
+  fextl::vector<Interval> Intervals; ///< list of intervals sorted by their end offset
+
+public:
+  struct QueryResult {
+    bool Enclosed; ///< If the given offset was enclosed by an interval
+    DifferenceType Size; ///< Size of the interval starting from the query offset, or distance to the next interval if
+                         /// `Enclosed` is false (if there is no next interval, size is 0)
+  };
+
+  void Clear() {
+    Intervals.clear();
+  }
+
+  void Insert(Interval Entry) {
+    auto FirstIt = std::upper_bound(Intervals.begin(), Intervals.end(), Entry, [](const auto &LHS, const auto &RHS) {
+      return LHS.Offset < RHS.End;
+    }); // Lowest offset interval that (maybe) overlaps with the new interval
+
+    if (FirstIt == Intervals.end() || FirstIt->Offset >= Entry.End) {
+      Intervals.insert(FirstIt, Entry);
```

Comment:
Does this handle the scenario `Intervals = { [0..10) }; Entry = [10..20)` correctly?

Have you considered adding some Catch2-based unit tests for this class?

Reply:
Yes it does, I considered unit tests but thought the code was probably small enough to be okay without

Reply:
It's not the most critical thing to unit tests indeed, but it's also very easy to write test code for. I would sleep better if we had coverage to ensure the edge cases are handled right :)